tornadox/blocks/sqrtutil.py

```diff
@@ -0,0 +1,99 @@
+"""Utility functions for square-root implementations."""
+
+from functools import partial
+
+import jax
+import jax.numpy as jnp
+import jax.scipy as jsp
+
+
+@jax.jit
+def correct_noisefree(*, h, c_sqrtm):
+    return correct_noisefree_matfree(h_mul_c_sqrtm=h @ c_sqrtm, c_sqrtm=c_sqrtm)
+
+
+@jax.jit
+def correct_noisefree_matfree(*, h_mul_c_sqrtm, c_sqrtm):
+    c_sqrtm_obs = sqrtm_to_cholesky(St=h_mul_c_sqrtm.T)
+    g = jsp.linalg.cho_solve((c_sqrtm_obs, True), h_mul_c_sqrtm @ c_sqrtm.T).T
+    c_sqrtm_cor = sqrtm_to_cholesky(St=(c_sqrtm - g @ h_mul_c_sqrtm).T)
+    return c_sqrtm_obs, (c_sqrtm_cor, g)
+
+
+@jax.jit
+@partial(jax.vmap, in_axes=(0, 0), out_axes=(0, (0, 0)))
+def correct_noisefree_matfree_batched(h_mul_c_sqrtm, c_sqrtm):
+    """Same as correct_noisefree_matfree, but batched."""
+    # careful: vmapped qr() (which this function calls)
+    # is not parallelised on GPU. See
+    # https://github.com/google/jax/issues/8542
+    return correct_noisefree_matfree(h_mul_c_sqrtm=h_mul_c_sqrtm, c_sqrtm=c_sqrtm)
+
+
+@jax.jit
+def correct_noisy(*, h, c_sqrtm, r_sqrtm):
+    return correct_noisy_matfree(
+        h_matmul_c_sqrtm=h @ c_sqrtm, c_sqrtm=c_sqrtm, r_sqrtm=r_sqrtm
+    )
+
+
+@jax.jit
+def correct_noisy_matfree(*, h_matmul_c_sqrtm, c_sqrtm, r_sqrtm):
+
+    output_dim, input_dim = h_matmul_c_sqrtm.shape
+
+    blockmat = jnp.block(
+        [
+            [r_sqrtm, h_matmul_c_sqrtm],
+            [jnp.zeros((input_dim, output_dim)), c_sqrtm],
+        ]
+    ).T
+
+    R = jnp.linalg.qr(blockmat, mode="r")
+
+    R1 = R[:output_dim, :output_dim]  # observed RV
+    R12 = R[:output_dim, output_dim:]  # something like the crosscov
+    R3 = R[output_dim:, output_dim:]  # corrected RV
+
+    # todo: what is going on here???
+    #  why lstsq? The matrix should be well-conditioned.
+    gain = jnp.linalg.lstsq(R1, R12)[0].T
+
+    c_sqrtm_cor = _make_diagonal_positive(L=R3.T)
+    c_sqrtm_obs = _make_diagonal_positive(L=R1.T)
+    return c_sqrtm_obs, (c_sqrtm_cor, gain)
+
+
+@jax.jit
+def sum_of_sqrtm_factors(*, S1, S2):
+    """Compute Cholesky factor of A @ SC @ SC.T @ A.T + SQ @ SQ.T"""
```

Comment:
I guess instead of A @ SC it should be S1, and S2 instead of SQ?

Reply:
yup